source/fab/steps/__init__.py

```diff
@@ -0,0 +1,70 @@
+import multiprocessing
+from typing import Dict
+
+
+# class Step(ABC):
+
+
+class Step(object):
+    """
+    Base class for build steps.
+
+    Provides multiprocessing capabilities which can be disabled for debugging.
+
+    """
+
+    def __init__(self, name):
+        self.name = name
+
+    # @abstractmethod
+    # todo: too much to send the whole config through?
+    #       it allows step manipulation, adding steps, etc
+    #       but it could let things get into a mess? discuss...
+    def run(self, artefacts: Dict, config):
```

Comment:
On the second point, you're right. `run` is not like `run_mp`. There was an old misleading comment stating otherwise which we've removed, sorry. We've also improved the documentation to explain that `run_mp` is a utility method which can be used in `run` to process artefacts in parallel.

A Step has a run method which consumes artefacts and creates new artefacts. That's common to all steps and seems like a no-brainer design.

I didn't understand the comment about _"run methods inheriting nature"_, sorry. Sounds like a topic for a discussion?